File: 2385-amount-of-time-for-binary-tree-to-be-infected/2385-amount-of-time-for-binary-tree-to-be-infected.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        up, q, st = {}, [root], None
        while q:
            l=len(q)
            for _ in range(l):
                node=q.pop(0)
                if node.val==start:
                    st=node
                if node.left:
                    q.append(node.left)
                    up[node.left.val]=node
                if node.right:
                    q.append(node.right)
                    up[node.right.val]=node
        q.append(st)
        ans, s = 0,set()
        while q:
            l=len(q)
            for _ in range(l):
                node=q.pop(0)
                s.add(node.val)
                if node.left and node.left.val not in s: q.append(node.left)
                if node.right and node.right.val not in s: q.append(node.right)
                if node.val in up and up[node.val].val not in s: q.append(up[node.val])
            ans+=1
        return ans-1

File: 2385-amount-of-time-for-binary-tree-to-be-infected/2385-amount-of-time-for-binary-tree-to-be-infected.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, TreeNode> par;
    int start;
    TreeNode startNode;
    public int amountOfTime(TreeNode root, int start) {
        par = new HashMap<>();
        this.start = start;
        dfs(root, null);
        return bfs(startNode);
    }
    
    void dfs(TreeNode root, TreeNode parent){
        if(root==null) return;
        par.put(root.val, parent);
        if(root.val == start) startNode = root;
        dfs(root.left, root);
        dfs(root.right, root);
    }
    
    int bfs(TreeNode start){
        Queue<TreeNode> q = new LinkedList<>();
        q.add(start);
        HashSet<Integer> vis = new HashSet<>();
        int lvl = 0;
        
        while(!q.isEmpty()){
            int l = q.size();
            
            for(int i=0; i<l; ++i){
                TreeNode node = q.poll();
                vis.add(node.val);
                
                if(node.left != null && !vis.contains(node.left.val))
                    q.offer(node.left);
                if(node.right != null &&  !vis.contains(node.right.val))
                    q.offer(node.right);
                if(par.get(node.val) != null && !vis.contains(par.get(node.val).val))
                    q.offer(par.get(node.val));
            }
            ++lvl;
        }
        
        return lvl-1;
    }
}